python/cudf_polars/cudf_polars/experimental/join.py

```diff
@@ -0,0 +1,275 @@
+        partitioned_on = ir.left_on
+    elif ir.options[0] == "right":  # pragma: no cover
+        partitioned_on = ir.right_on
+    partition_info[new_node] = PartitionInfo(
+        count=output_count,
+        partitioned_on=partitioned_on,
+    )
+
+    return new_node, partition_info
+
+
+def _make_bcast_join(
+    ir: Join,
+    output_count: int,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    left: IR,
+    right: IR,
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    how = ir.options[0]
+    if how != "inner":
+        shuffle_options: dict[str, Any] = {}
+        left_count = partition_info[left].count
+        right_count = partition_info[right].count
+        if left_count >= right_count:
+            right = Shuffle(
+                right.schema,
+                ir.right_on,
+                shuffle_options,
+                right,
+            )
+            partition_info[right] = PartitionInfo(
+                count=right_count,
+                partitioned_on=ir.right_on,
+            )
+        else:
+            left = Shuffle(
+                left.schema,
+                ir.left_on,
+                shuffle_options,
+                left,
+            )
+            partition_info[left] = PartitionInfo(
+                count=left_count,
+                partitioned_on=ir.left_on,
+            )
+
+    new_node = ir.reconstruct([left, right])
+    partition_info[new_node] = PartitionInfo(count=output_count)
+    return new_node, partition_info
+
+
+@lower_ir_node.register(Join)
+def _(
+    ir: Join, rec: LowerIRTransformer
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    # Lower children
+    children, _partition_info = zip(*(rec(c) for c in ir.children), strict=True)
+    partition_info = reduce(operator.or_, _partition_info)
+
+    how = ir.options[0]
+    left, right = children
+    output_count = max(partition_info[left].count, partition_info[right].count)
+    if output_count == 1:
+        new_node = ir.reconstruct(children)
+        partition_info[new_node] = PartitionInfo(count=1)
+        return new_node, partition_info
+    elif how == "cross":
+        raise NotImplementedError(
+            "cross join not support for multiple partitions."
+        )  # pragma: no cover
+
+    # Use a broadcast join if appropriate
+    if partition_info[left].count >= partition_info[right].count:
+        bcast_count = partition_info[right].count
+        other = left
+        other_on = ir.left_on
+    else:
+        bcast_count = partition_info[left].count
+        other = right
+        other_on = ir.right_on
+
+    # Avoid the broadcast if the "large" table is already shuffled
+    other_shuffled = (
+        partition_info[other].partitioned_on == other_on
+        and partition_info[other].count == output_count
+    )
+
+    if (
+        not other_shuffled
+        and bcast_count <= 8  # TODO: Make this configurable
```

Comment:
For now we arbitrarily switch from a broadcast to a shuffle when the "small" side has >8 partitions. I don't think we need to settle on the best heuristic in this particular PR (we just want Join support).

Reply:
It might be worth adding that in the TODO comment.